File: src/main/java/org/voegtle/weatherwidget/widget/SmallWidgetUpdateTask.java
package org.voegtle.weatherwidget.widget;

import android.content.Context;
import android.util.Log;
import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.preferences.ApplicationSettings;

public class SmallWidgetUpdateTask extends AbstractWidgetUpdateTask<String, Void, WeatherData> {

  private final SmallWidgetScreenPainter screenPainter;

  public SmallWidgetUpdateTask(Context context, ApplicationSettings configuration,
                               SmallWidgetScreenPainter screenPainter) {
    super(context, configuration);
    this.screenPainter = screenPainter;
  }

  @Override
  protected void onPreExecute() {
    screenPainter.showDataIsInvalid();
    screenPainter.updateAllWidgets();
  }


  @Override
  protected WeatherData doInBackground(String... weatherServerUrl) {
    return fetchWeatherData(weatherServerUrl[0]);
  }

  @Override
  protected void onPostExecute(WeatherData data) {
    try {
      screenPainter.updateData(data);
    } catch (Throwable th) {
      Log.e(SmallWidgetUpdateTask.class.toString(), "Failed to update View", th);
    } finally {
      screenPainter.showDataIsValid();
    }

  }
}

File: src/main/java/org/voegtle/weatherwidget/widget/SmallWidgetUpdateTask.kt
package org.voegtle.weatherwidget.widget

import android.content.Context
import android.util.Log
import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.preferences.ApplicationSettings

class SmallWidgetUpdateTask(context: Context, configuration: ApplicationSettings,
                            private val screenPainter: SmallWidgetScreenPainter)
  : AbstractWidgetUpdateTask<String, Void, WeatherData>(context, configuration) {

  override fun onPreExecute() {
    screenPainter.showDataIsInvalid()
    screenPainter.updateAllWidgets()
  }


  override fun doInBackground(vararg weatherServerUrl: String): WeatherData? {
    return fetchWeatherData(weatherServerUrl[0])
  }

  override fun onPostExecute(data: WeatherData?) {
    try {
      screenPainter.updateData(data)
    } catch (th: Throwable) {
      Log.e(SmallWidgetUpdateTask::class.java.toString(), "Failed to update View", th)
    } finally {
      screenPainter.showDataIsValid()
    }

  }
}
